#**스마트폰 센서 데이터 기반 모션 분류**
# 단계3 : 단계별 모델링


## 0.미션

* 나만의 앙상블 모델 만들기.
    * 1. 다양한 종류의 모델을 생성(미션3에서 생성한 모델 활용)
    * 2. 각 모델을 통합하여 예측하는 앙상블 모델링
        * Voting, Stacking
        * 앙상블을 위한 다양한 시도를 해 봅시다.
            - 예① : 미션3에서 생성한 모델을 이용한 앙상블
            - 예② : Random Jungle!(Random Forest 여러 개로 예측하는 모델)

* 파이프라인 구축
    * 테스트 데이터를 입력하면, 전처리 + 예측 예측결과가 나오도록 파이프라인을 구성합니다.

* 성능 가이드
    * Accuracy : 0.97 ~


## 1.환경설정

* 세부 요구사항
    - 경로 설정 : 로컬 수행(Ananconda)
        * 제공된 압축파일을 다운받아 압축을 풀고
        * anaconda의 root directory(보통 C:/Users/< ID > 에 project3_1 폴더를 만들고, 복사해 넣습니다.
    - 기본적으로 필요한 라이브러리를 import 하도록 코드가 작성되어 있습니다.
        * 필요하다고 판단되는 라이브러리를 추가하세요.


### (1) 라이브러리 로딩

In [125]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm.auto import tqdm
import warnings
warnings.filterwarnings('ignore')

import joblib
from sklearn.ensemble import StackingClassifier, VotingClassifier

# 필요한 라이브러리, 함수 로딩 ------------------
from sklearn.model_selection import train_test_split
from sklearn.metrics import *
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from lightgbm import LGBMClassifier

* 제공 함수 생성
    * 변수 중요도를 시각화할 수 있는 함수를 제공합니다.
    * 입력 :
        * importance : 트리모델의 변수 중요도(예: model.feature_importances_)
        * names : 변수 이름 목록(예 : x_train.columns
        * result_only  : 변수 중요도 순으로 데이터프레임만 return할지, 그래프도 포함할지 결정. False이면 결과 데이터프레임 + 그래프
        * topn : 중요도 상위 n개만 표시. all 이면 전체.
    * 출력 :
        * 중요도 그래프 : 중요도 내림차순으로 정렬
        * 중요도 데이터프레임 : 중요도 내림차순으로 정렬

In [14]:
# 변수의 특성 중요도 계산하기
def plot_feature_importance(importance, names, result_only = False, topn = 'all'):
    feature_importance = np.array(importance)
    feature_name = np.array(names)

    data={'feature_name':feature_name,'feature_importance':feature_importance}
    fi_temp = pd.DataFrame(data)

    #변수의 특성 중요도 순으로 정렬하기
    fi_temp.sort_values(by=['feature_importance'], ascending=False,inplace=True)
    fi_temp.reset_index(drop=True, inplace = True)

    if topn == 'all' :
        fi_df = fi_temp.copy()
    else :
        fi_df = fi_temp.iloc[:topn]

    #변수의 특성 중요도 그래프로 그리기
    if result_only == False :
        plt.figure(figsize=(10,20))
        sns.barplot(x='feature_importance', y='feature_name', data = fi_df)

        plt.xlabel('importance')
        plt.ylabel('feature name')
        plt.grid()

    return fi_df

### (2) 데이터 불러오기

* 주어진 데이터셋
    * data01_train.csv : 학습 및 검증용
    * data01_test.csv : 테스트용
    
* 세부 요구사항
    * 칼럼 삭제 : data01_train.csv와 data01_test.csv 에서 'subject' 칼럼은 불필요하므로 삭제합니다.

#### 1) 데이터로딩

In [19]:
file1 = 'data01_train.csv'
file2 = 'data01_test.csv'

In [21]:
data = pd.read_csv(file1)
test = pd.read_csv(file2)

In [22]:
# 불필요한 칼럼 삭제
data.drop('subject', axis=1, inplace=True)
test.drop('subject', axis=1, inplace=True)

#### 2) 기본 정보 조회

In [26]:
#전체 데이터의 행,열 개수 확인
data.shape

(5881, 562)

In [28]:
#전체 데이터의 상위 5개 행 확인
data.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288508          -0.009196          -0.103362         -0.988986   
1           0.265757          -0.016576          -0.098163         -0.989551   
2           0.278709          -0.014511          -0.108717         -0.997720   
3           0.289795          -0.035536          -0.150354         -0.231727   
4           0.394807           0.034098           0.091229          0.088489   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.962797         -0.967422         -0.989000         -0.962596   
1         -0.994636         -0.987435         -0.990189         -0.993870   
2         -0.981088         -0.994008         -0.997934         -0.982187   
3         -0.006412         -0.338117         -0.273557          0.014245   
4         -0.106636         -0.388502         -0.010469         -0.109680   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-skewness()  \
0         -0.965650         -0.929747  ...                        -0.487737   
1         -0.987558         -0.937337  ...                        -0.237820   
2         -0.995017         -0.942584  ...                        -0.535287   
3         -0.347916          0.008288  ...                        -0.004012   
4         -0.346372          0.584131  ...                        -0.157832   

   fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
0                        -0.816696                    -0.042494   
1                        -0.693515                    -0.062899   
2                        -0.829311                     0.000265   
3                        -0.408956                    -0.255125   
4                        -0.563437                    -0.044344   

   angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
0                             -0.044218                          0.307873   
1                              0.388459                         -0.765014   
2                             -0.525022                         -0.891875   
3                              0.612804                          0.747381   
4                             -0.845268                         -0.974650   

   angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
0                              0.072790             -0.601120   
1                              0.771524              0.345205   
2                              0.021528             -0.833564   
3                             -0.072944             -0.695819   
4                             -0.887846             -0.705029   

   angle(Y,gravityMean)  angle(Z,gravityMean)            Activity  
0              0.331298              0.165163            STANDING  
1             -0.769186             -0.147944              LAYING  
2              0.202434             -0.032755            STANDING  
3              0.287154              0.111388             WALKING  
4              0.264952              0.137758  WALKING_DOWNSTAIRS  

[5 rows x 562 columns]

In [30]:
#전체 데이터의 수치형 변수 분포 확인
data.describe()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
count        5881.000000        5881.000000        5881.000000   
mean            0.274811          -0.017799          -0.109396   
std             0.067614           0.039422           0.058373   
min            -0.503823          -0.684893          -1.000000   
25%             0.262919          -0.024877          -0.121051   
50%             0.277154          -0.017221          -0.108781   
75%             0.288526          -0.010920          -0.098163   
max             1.000000           1.000000           1.000000   

       tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
count       5881.000000       5881.000000       5881.000000       5881.000000   
mean          -0.603138         -0.509815         -0.604058         -0.628151   
std            0.448807          0.501815          0.417319          0.424345   
min           -1.000000         -0.999844         -0.999667         -1.000000   
25%           -0.992774         -0.977680         -0.980127         -0.993602   
50%           -0.943933         -0.844575         -0.856352         -0.948501   
75%           -0.242130         -0.034499         -0.262690         -0.291138   
max            1.000000          0.916238          1.000000          1.000000   

       tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
count       5881.000000       5881.000000       5881.000000  ...   
mean          -0.525944         -0.605374         -0.465490  ...   
std            0.485115          0.413043          0.544995  ...   
min           -0.999419         -1.000000         -1.000000  ...   
25%           -0.977865         -0.980112         -0.936067  ...   
50%           -0.849266         -0.849896         -0.878729  ...   
75%           -0.068857         -0.268539         -0.013690  ...   
max            0.967664          1.000000          1.000000  ...   

       fBodyBodyGyroJerkMag-meanFreq()  fBodyBodyGyroJerkMag-skewness()  \
count                      5881.000000                      5881.000000   
mean                          0.126955                        -0.305883   
std                           0.249176                         0.322808   
min                          -0.965725                        -0.979261   
25%                          -0.021610                        -0.541969   
50%                           0.133887                        -0.342923   
75%                           0.288944                        -0.127371   
max                           0.946700                         0.989538   

       fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
count                      5881.000000                  5881.000000   
mean                         -0.623548                     0.008524   
std                           0.310371                     0.339730   
min                          -0.999765                    -0.976580   
25%                          -0.845985                    -0.122361   
50%                          -0.712677                     0.010278   
75%                          -0.501158                     0.154985   
max                           0.956845                     1.000000   

       angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
count                           5881.000000                       5881.000000   
mean                              -0.001185                          0.009340   
std                                0.447197                          0.608190   
min                               -1.000000                         -1.000000   
25%                               -0.294369                         -0.481718   
50%                                0.005146                          0.011448   
75%                                0.285030                          0.499857   
max                                1.000000                          0.998702   

       angle(tBodyGyroJ

In [31]:
#전체 데이터의 모든 변수 확인
data.columns

Index(['tBodyAcc-mean()-X', 'tBodyAcc-mean()-Y', 'tBodyAcc-mean()-Z',
       'tBodyAcc-std()-X', 'tBodyAcc-std()-Y', 'tBodyAcc-std()-Z',
       'tBodyAcc-mad()-X', 'tBodyAcc-mad()-Y', 'tBodyAcc-mad()-Z',
       'tBodyAcc-max()-X',
       ...
       'fBodyBodyGyroJerkMag-skewness()', 'fBodyBodyGyroJerkMag-kurtosis()',
       'angle(tBodyAccMean,gravity)', 'angle(tBodyAccJerkMean),gravityMean)',
       'angle(tBodyGyroMean,gravityMean)',
       'angle(tBodyGyroJerkMean,gravityMean)', 'angle(X,gravityMean)',
       'angle(Y,gravityMean)', 'angle(Z,gravityMean)', 'Activity'],
      dtype='object', length=562)

## **2. 데이터 전처리**

* 가변수화, 데이터 분할, NaN 확인 및 조치, 스케일링 등 필요한 전처리를 수행한다.


### (1) 데이터 분할1 : x, y

* 세부 요구사항
    - x, y로 분할합니다.

In [34]:
drop_cols ='Activity'
x = data.drop(columns = drop_cols)
y = data.loc[:,drop_cols]

x_test = test.drop(columns = drop_cols)
y_test= test.loc[:,drop_cols]

### (2) 데이터분할2 : train, validation

* 세부 요구사항
    - train : val = 8 : 2 혹은 7 : 3
    - random_state 옵션을 사용하여 다른 모델과 비교를 위해 성능이 재현되도록 합니다.

In [36]:
x_train , x_val , y_train , y_val = train_test_split(x,y,test_size = 0.3)

### (3) 스케일링


* 세부 요구사항
    - 스케일링을 필요로 하는 알고리즘 사용을 위해서 코드 수행
    - min-max 방식 혹은 standard 방식 중 한가지 사용.

In [38]:
scaler = MinMaxScaler()
scaler.fit(x_train)
x_train = scaler.transform(x_train)
x_val = scaler.transform(x_val)
x_test = scaler.transform(x_test)

### (4) (옵션)KNN을 위한 전처리
* KNN은 대량의 데이터에 대한 거리 계산을 할때, 데이터를 연속적인 메모리 블록에 저장하는 것이 중요합니다.
* c_contiguous 속성을 요구

In [41]:
x_train = np.ascontiguousarray(x_train)
x_val = np.ascontiguousarray(x_val)

### (5) (옵션) Activity 범주 0~5 변경

In [43]:
activity_mapping = {
    'LAYING': 0,
    'STANDING': 1,
    'SITTING': 2,
    'WALKING': 3,
    'WALKING_UPSTAIRS': 4,
    'WALKING_DOWNSTAIRS': 5
}
y_train = y_train.map(activity_mapping)
y_val = y_val.map(activity_mapping)
y_test = y_test.map(activity_mapping)

## **3.앙상블 모델링**
* Voting 혹은 Stacking 방식의 앙상블 모델을 구성하고 성능을 평가해 봅시다.


### (1) 모델 로딩하기

* 세부 요구사항
    * 2.기본 모델링 파일에서 저장한 모델을 로딩합니다.(joblib.load)


In [9]:
import joblib

model_knn = joblib.load('knn.pkl')
model_lgbm = joblib.load('LGBM.pkl')
model_lr = joblib.load('lr.pkl')
model_svm = joblib.load('svm.pkl')

### (2) 앙상블 모델 생성


#### Voting

In [85]:
voting_clf = VotingClassifier(estimators=[
    ('knn', model_knn), 
    ('svm', model_svm), 
    ('lr', model_lr), 
    ('lgbm', model_lgbm)], voting='hard', weights=[4, 1, 2, 3])  # 가중치(전 파일에서 정확도가 더 잘 나온 순으로 지정했습니다!)

In [87]:
# 모델 학습
voting_clf.fit(x_train, y_train)

VotingClassifier(estimators=[('knn',
                              KNeighborsClassifier(metric='manhattan',
                                                   n_neighbors=3,
                                                   weights='distance')),
                             ('svm', SVC(C=10, gamma=0.1)),
                             ('lr',
                              LogisticRegression(C=10, solver='newton-cg')),
                             ('lgbm',
                              GridSearchCV(cv=5,
                                           estimator=LGBMClassifier(verbose=-1),
                                           param_grid={'learning_rate': [0.422],
                                                       'n_estimators': [260]}))],
                 weights=[4, 1, 2, 3])

In [91]:
# 평가
y_pred = voting_clf.predict(x_test)

# 결과
print(confusion_matrix(y_test, y_pred))
print()
print('Classification Report  \n' , classification_report(y_test, y_pred))

report = classification_report(y_test, y_pred, output_dict=True)
accuracy = report['accuracy']
print(f'Accuracy: {accuracy:.4f}')

[[292   0   0   0   0   0]
 [  0 282   5   0   0   0]
 [  0   8 246   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report  
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.97      0.98      0.98       287
           2       0.98      0.97      0.97       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9912


#### Stacking

##### Logistic Regression: 간단하고 빠름

In [95]:
# 최종 모델 Logistic Regression
stacking_clf_lr = StackingClassifier(estimators=[
    ('knn', model_knn), 
    ('svm', model_svm), 
    ('lr', model_lr), 
    ('lgbm', model_lgbm)],
    final_estimator=LogisticRegression(), cv=5
)

In [97]:
# 모델 학습
stacking_clf_lr.fit(x_train, y_train)

StackingClassifier(cv=5,
                   estimators=[('knn',
                                KNeighborsClassifier(metric='manhattan',
                                                     n_neighbors=3,
                                                     weights='distance')),
                               ('svm', SVC(C=10, gamma=0.1)),
                               ('lr',
                                LogisticRegression(C=10, solver='newton-cg')),
                               ('lgbm',
                                GridSearchCV(cv=5,
                                             estimator=LGBMClassifier(verbose=-1),
                                             param_grid={'learning_rate': [0.422],
                                                         'n_estimators': [260]}))],
                   final_estimator=LogisticRegression())

In [108]:
# 평가
y_pred_stacking_lr = stacking_clf_lr.predict(x_test)

# 결과
print(confusion_matrix(y_test, y_pred_stacking_lr))
print()
print('Classification Report  \n' , classification_report(y_test, y_pred_stacking_lr))

report_lr = classification_report(y_test, y_pred_stacking_lr, output_dict=True)
accuracy_lr = report_lr['accuracy']
print(f'Accuracy: {accuracy_lr:.4f}')

[[292   0   0   0   0   0]
 [  0 284   3   0   0   0]
 [  0   7 247   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report  
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.98      0.99      0.98       287
           2       0.99      0.97      0.98       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9932


##### Random Forest: 복잡한 의사결정을 다루고, 다양한 특성 처리에 용이

In [112]:
# 최종 모델 Random Forest 사용
stacking_clf_rf = StackingClassifier(estimators=[
        ('knn', model_knn), 
        ('svm', model_svm), 
        ('lr', model_lr), 
        ('lgbm', model_lgbm)],
    final_estimator=RandomForestClassifier(), cv=5
)

In [114]:
# 모델 학습
stacking_clf_rf.fit(x_train, y_train)

StackingClassifier(cv=5,
                   estimators=[('knn',
                                KNeighborsClassifier(metric='manhattan',
                                                     n_neighbors=3,
                                                     weights='distance')),
                               ('svm', SVC(C=10, gamma=0.1)),
                               ('lr',
                                LogisticRegression(C=10, solver='newton-cg')),
                               ('lgbm',
                                GridSearchCV(cv=5,
                                             estimator=LGBMClassifier(verbose=-1),
                                             param_grid={'learning_rate': [0.422],
                                                         'n_estimators': [260]}))],
                   final_estimator=RandomForestClassifier())

In [115]:
# 평가
y_pred_stacking_rf = stacking_clf_rf.predict(x_test)

# 결과
print(confusion_matrix(y_test, y_pred_stacking_rf))
print()
print('Classification Report  \n' , classification_report(y_test, y_pred_stacking_rf))

report_rf = classification_report(y_test, y_pred_stacking_rf, output_dict=True)
accuracy_rf = report_rf['accuracy']
print(f'Accuracy: {accuracy_rf:.4f}')

[[292   0   0   0   0   0]
 [  0 284   3   0   0   0]
 [  0   9 245   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report  
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.97      0.99      0.98       287
           2       0.99      0.96      0.98       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9918


##### SVM:고차원 데이터에 적합

In [127]:
# 최종 모델 SVM 사용
stacking_clf_svm = StackingClassifier(estimators=[
        ('knn', model_knn), 
        ('svm', model_svm), 
        ('lr', model_lr), 
        ('lgbm', model_lgbm)],
    final_estimator=SVC(), cv=5
)

In [129]:
# 모델 학습
stacking_clf_svm.fit(x_train, y_train)

StackingClassifier(cv=5,
                   estimators=[('knn',
                                KNeighborsClassifier(metric='manhattan',
                                                     n_neighbors=3,
                                                     weights='distance')),
                               ('svm', SVC(C=10, gamma=0.1)),
                               ('lr',
                                LogisticRegression(C=10, solver='newton-cg')),
                               ('lgbm',
                                GridSearchCV(cv=5,
                                             estimator=LGBMClassifier(verbose=-1),
                                             param_grid={'learning_rate': [0.422],
                                                         'n_estimators': [260]}))],
                   final_estimator=SVC())

In [130]:
# 평가
y_pred_stacking_svm = stacking_clf_svm.predict(x_test)

# 결과
print(confusion_matrix(y_test, y_pred_stacking_svm))
print()
print('Classification Report  \n' , classification_report(y_test, y_pred_stacking_svm))

report_svm = classification_report(y_test, y_pred_stacking_svm, output_dict=True)
accuracy_svm = report_svm['accuracy']
print(f'Accuracy: {accuracy_svm:.4f}')

[[292   0   0   0   0   0]
 [  0 284   3   0   0   0]
 [  0   8 246   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report  
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.97      0.99      0.98       287
           2       0.99      0.97      0.98       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9925


##### LGBM: 복잡한 데이터에 적합

In [135]:
# 최종 모델 LightGBM사용
stacking_clf_lgbm = StackingClassifier(estimators=[
        ('knn', model_knn), 
        ('svm', model_svm), 
        ('lr', model_lr), 
        ('lgbm', model_lgbm)],
    final_estimator=LGBMClassifier(), cv=5
)

In [137]:
# 모델 학습
stacking_clf_lgbm.fit(x_train, y_train)

StackingClassifier(cv=5,
                   estimators=[('knn',
                                KNeighborsClassifier(metric='manhattan',
                                                     n_neighbors=3,
                                                     weights='distance')),
                               ('svm', SVC(C=10, gamma=0.1)),
                               ('lr',
                                LogisticRegression(C=10, solver='newton-cg')),
                               ('lgbm',
                                GridSearchCV(cv=5,
                                             estimator=LGBMClassifier(verbose=-1),
                                             param_grid={'learning_rate': [0.422],
                                                         'n_estimators': [260]}))],
                   final_estimator=LGBMClassifier())

In [138]:
# 평가
y_pred_stacking_lgbm = stacking_clf_lgbm.predict(x_test)

# 결과
print(confusion_matrix(y_test, y_pred_stacking_lgbm))
print()
print('Classification Report  \n' , classification_report(y_test, y_pred_stacking_lgbm))

report_lgbm = classification_report(y_test, y_pred_stacking_lgbm, output_dict=True)
accuracy_lgbm = report_lgbm['accuracy']
print(f'Accuracy: {accuracy_lgbm:.4f}')

[[292   0   0   0   0   0]
 [  0 285   2   0   0   0]
 [  0   8 246   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report  
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.97      0.99      0.98       287
           2       0.99      0.97      0.98       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9932


---
- 최종 모델로 Logistic Regression와 LightGBM을 사용했을 때 0.9932로 가장 높은 정확도를 보임
- 하지만 학습 시간이 Logistic Regression이 훨씬 빨라서 성능이 같다면 이걸 사용하는게 좋을듯!

## 4.파이프라인 구성

### (1) 함수 만들기

In [150]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ('scaler', MinMaxScaler()),    # MinMaxScaler로 전처리
    ('stacking_clf_lr', stacking_clf_lr) # Stacking 앙상블 모델
])

In [154]:
# 파이프라인 학습
pipeline.fit(x_train, y_train)

Pipeline(steps=[('scaler', MinMaxScaler()),
                ('stacking_clf_lr',
                 StackingClassifier(cv=5,
                                    estimators=[('knn',
                                                 KNeighborsClassifier(metric='manhattan',
                                                                      n_neighbors=3,
                                                                      weights='distance')),
                                                ('svm', SVC(C=10, gamma=0.1)),
                                                ('lr',
                                                 LogisticRegression(C=10,
                                                                    solver='newton-cg')),
                                                ('lgbm',
                                                 GridSearchCV(cv=5,
                                                              estimator=LGBMClassifier(verbose=-1),
                                                              param_grid={'learning_rate': [0.422],
                                                                          'n_estimators': [260]}))],
                                    final_estimator=LogisticRegression()))])

### (2) test 셋으로 예측하고 평가하기

In [155]:
# 예측
predictions = pipeline.predict(x_test)

In [160]:
# 결과
print('Confusion Matrix: \n', confusion_matrix(y_test, predictions))
print()
print('Classification Report \n' , classification_report(y_test, predictions))

report_lgbm = classification_report(y_test, y_pred_stacking_lgbm, output_dict=True)
accuracy_lgbm = report_lgbm['accuracy']
print(f'Accuracy: {accuracy_lgbm:.4f}')

Confusion Matrix: 
 [[292   0   0   0   0   0]
 [  0 284   3   0   0   0]
 [  0   7 247   0   0   0]
 [  0   0   0 228   0   0]
 [  0   0   0   0 215   0]
 [  0   0   0   0   0 195]]

Classification Report 
               precision    recall  f1-score   support

           0       1.00      1.00      1.00       292
           1       0.98      0.99      0.98       287
           2       0.99      0.97      0.98       254
           3       1.00      1.00      1.00       228
           4       1.00      1.00      1.00       215
           5       1.00      1.00      1.00       195

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Accuracy: 0.9932
